### PR TITLE
Fix bugs related to section reductions.

### PR DESCRIPTION
**Pull request by ZwFink**
As described in #180, reductions not do always work when the chares are contributing within a section. This pull request fixes the issue and provides tests for the new behavior.

### CODE REVIEW

**Review by rbuch — COMMENTED** [review]
This should all probably be run through some kind of style checker (`flake8` or something like that), there are some formatting inconsistencies (plus we should be running those kinda tools anyway).

File: charm4py/reduction.py [hunk 1]
```diff
@@ -34,26 +35,95 @@ class number:
...
+
+    try:
+        return min(contribs)
+    # Numpy can throw ValueError
+    except TypeError:
+        return _pairwiseOp(min, contribs)
+
+def _and(contribs):
+    if _useNumpyForReduction(contribs):
+        return np.logical_and.reduce(contribs)
```
[the remark this answers is not in the document]
Comment (ZwFink) [reply charm4py/reduction.py:97]: I think doing this in another PR is appropriate, but definitely it is a good idea.

File: charm4py/reduction.py [hunk 2]
```diff
@@ -34,26 +35,95 @@ class number:
...
+    except TypeError:
+        return _elementwiseOp(lambda x,y: x or y, contribs)
+
+def _xor(contribs):
+    if _useNumpyForReduction(contribs):
+        return np.logical_xor.reduce(contribs)
+
+    try:
+        iter(contribs[0])
+        return _pairwiseOp(lambda x,y: x ^ y, contribs)
```
[the remark this answers is not in the document]
Comment (ZwFink) [reply charm4py/reduction.py]: Good point. In the logical reductions x and y being bool is a precondition enforced by the runtime system, but I have added the cast to make it explicit.

[the remark this answers is not in the document]
Comment (ZwFink) [reply charm4py/reduction.py]: Regarding flake8: Is there a specific config that should be used, or is conforming to the default flake8 standard enough?

[the remark this answers is not in the document]
Comment (ZwFink) [reply charm4py/reduction.py]: @rbuch Just want to follow up on your thoughts on flake8

[the remark this answers is not in the document]
Comment (rbuch) [reply charm4py/reduction.py]: Sorry, I missed this earlier. I'm not terribly too well versed in it, so I'd say just big standard flake 8 is probably good enough.

**Review by rbuch — APPROVED** [review]

**Review by minitu — APPROVED** [review]

**Pull request merged**